client/network/src/discovery.rs

```diff
@@ -230,12 +230,18 @@ pub struct DiscoveryBehaviour {
 
 impl DiscoveryBehaviour {
 	/// Returns the list of nodes that we know exist in the network.
-	pub fn known_peers(&mut self) -> impl Iterator<Item = &PeerId> {
-		let mut set = HashSet::new();
-		for p in self.kademlias.values_mut().map(|k| k.kbuckets_entries()).flatten() {
-			set.insert(p);
+	pub fn known_peers(&mut self) -> HashSet<PeerId> {
```

Comment:
What is the motivation for the switch to `HashSet` here, instead of just switching from

```rust
impl Iterator<Item = &PeerId>
```

to

```rust
impl Iterator<Item = PeerId>
```
?

Reply:
Returning a `HashSet` implies that the collection returned by the method does not contain any duplication whereas an `Iterator` does not give such guarantee.

I don't think it has any performance implications whether the method calls `into_iter` or the consumer calls `into_iter`.

That said I am fine changing this to return an `Iterator` instead. What would you prefer @twittner?

Reply:
If you want to commit to set semantics feel free to change it. I was just curious if there was a specific reason for this.

Reply:
Since the suggestion came from me, I am curious. I tend to think that when a function builds an owned collection and then only returns an iterator, there is an unfortunate loss of structure and very often consuming code then eventually `collect()`s again. But maybe if someone were to call `known_peers().collect::<HashSet<_>>()` with `known_peers` return `impl Iterator` the compiler would optimise the `HashSet::into_iter() -> collect::<HashSet<_>>` away?

Reply:
I assume you are worried about performance, not so much about returning an existential type. I do not know though what — if any — compiler optimisations apply here.

Reply:
> I do not know though what — if any — compiler optimisations apply here.

As far as I can tell `known_peers` is only used within `NetworkWorker::network_state` which specifically states:

> Use this only for debugging.

That said I am still curious and thus wrote a [small benchmark](https://github.com/mxinden/into-iter-collect-optimization). As far as I can tell Rustc or LLVM can not optimize the `HashSet::into_iter() -> collect::<HashSet<_>>` case. @romanb @twittner I would be curious if you can spot something I am missing.

Reply:
@mxinden I don't think there are any such optimisations one could rely upon. It just seems that there are some such "in-place" optimisations for specific collections and iterator chains (see e.g. https://github.com/rust-lang/rust/pull/70793).